### ISSUE TITLE
hwloc and embedded mode

### ISSUE THREAD

**Issue by ggouaillardet** [pmix/pmix#1045]
If I build Open MPI with the embedded `hwloc` and `PMIx`, the (mangled) `hwloc` symbols provided by `libopen-pal.so` are used (for example in `mca_pnet_opa.so`) __but__ this pmix plugin is linked with the system `libhwloc.so`.

The reason is when `hwloc` is detected, `-lhwloc` is appended to `LIBS` by `config/pmix_setup_hwloc.m4`.

`config/pmix_setup_libevent.m4` assumes everything is there when building the embedded mode (which is fine since `libevent` is a requirement of PMIx).
The very same logic cannot be applied here since `hwloc` is optional by default.

I ended up with this patch to give Open MPI a hacky way to do nothing and have PMIx assume `hwloc` is there and functional (unlike in PMIx, `hwloc` is a requirement of Open MPI)

```patch
diff --git a/opal/mca/pmix/pmix4x/configure.m4 b/opal/mca/pmix/pmix4x/configure.m4
index 9b84e73..49b16d6 100644
--- a/opal/mca/pmix/pmix4x/configure.m4
+++ b/opal/mca/pmix/pmix4x/configure.m4
@@ -60,7 +60,7 @@ AC_DEFUN([MCA_opal_pmix_pmix4x_CONFIG],[
           [opal_pmix_pmix4x_args="--disable-debug $opal_pmix_pmix4x_args"
            CFLAGS="$OPAL_CFLAGS_BEFORE_PICKY $OPAL_VISIBILITY_CFLAGS"])
     AC_MSG_CHECKING([if want to install standalone libpmix])
-    AS_IF([test "$enable_install_libpmix" == "yes"],
+    AS_IF([test "$enable_install_libpmix" = "yes"],
           [AC_MSG_RESULT([yes])],
           [AC_MSG_RESULT([no])
            opal_pmix_pmix4x_args="--with-pmix-symbol-rename=OPAL_MCA_PMIX4X_ --enable-embedded-mode $opal_pmix_pmix4x_args"])
@@ -68,8 +68,11 @@ AC_DEFUN([MCA_opal_pmix_pmix4x_CONFIG],[
           [opal_pmix_pmix4x_args="--with-devel-headers  $opal_pmix_pmix4x_args"])
     CPPFLAGS="-I$OPAL_TOP_SRCDIR -I$OPAL_TOP_BUILDDIR -I$OPAL_TOP_SRCDIR/opal/include -I$OPAL_TOP_BUILDDIR/opal/include $CPPFLAGS"
 
+    AS_IF([test "$opal_hwloc_external_support" != "yes"],
+          [opal_pmix_pmix4x_hwloc=\'pmix_cv_hwloc_support=1\'])
+
     OPAL_CONFIG_SUBDIR([$opal_pmix_pmix4x_basedir/pmix],
-                       [$opal_pmix_pmix4x_args $opal_subdir_args 'CFLAGS=$CFLAGS' 'CPPFLAGS=$CPPFLAGS'],
+                       [$opal_pmix_pmix4x_args $opal_subdir_args $opal_pmix_pmix4x_hwloc 'CFLAGS=$CFLAGS' 'CPPFLAGS=$CPPFLAGS'],
                        [opal_pmix_pmix4x_happy=1], [opal_pmix_pmix4x_happy=0])
 
     CFLAGS=$opal_pmix_pmix4x_save_CFLAGS
diff --git a/opal/mca/pmix/pmix4x/pmix/config/pmix_setup_hwloc.m4 b/opal/mca/pmix/pmix4x/pmix/config/pmix_setup_hwloc.m4
index 86a9b9d..a2e0dde 100644
--- a/opal/mca/pmix/pmix4x/pmix/config/pmix_setup_hwloc.m4
+++ b/opal/mca/pmix/pmix4x/pmix/config/pmix_setup_hwloc.m4
@@ -23,12 +23,11 @@ AC_DEFUN([PMIX_HWLOC_CONFIG],[
                 [AC_HELP_STRING([--with-hwloc-libdir=DIR],
                                 [Search for hwloc libraries in DIR ])])
 
-    pmix_hwloc_support=0
     AS_IF([test "$with_hwloc" = "internal" || test "$with_hwloc" = "external"],
           [with_hwloc=])
 
     if test "$with_hwloc" != "no"; then
-        AC_MSG_CHECKING([for hwloc in])
+      AC_CACHE_CHECK([for hwloc in], [pmix_cv_hwloc_support],[
         if test ! -z "$with_hwloc" && test "$with_hwloc" != "yes"; then
             pmix_hwloc_dir=$with_hwloc
             pmix_hwloc_standard_header_location=no
@@ -60,9 +59,9 @@ AC_DEFUN([PMIX_HWLOC_CONFIG],[
                            [-lhwloc],
                            [$pmix_hwloc_dir],
                            [$pmix_hwloc_libdir],
-                           [pmix_hwloc_support=1],
-                           [pmix_hwloc_support=0])
-        if test $pmix_hwloc_support = "1"; then
+                           [pmix_cv_hwloc_support=1],
+                           [pmix_cv_hwloc_support=0])
+        if test $pmix_cv_hwloc_support = "1"; then
             LIBS="$LIBS -lhwloc"
             PMIX_EMBEDDED_LIBS="$PMIX_EMBEDDED_LIBS -lhwloc"
             if test "$pmix_hwloc_standard_header_location" != "yes"; then
@@ -74,14 +73,17 @@ AC_DEFUN([PMIX_HWLOC_CONFIG],[
                 LDFLAGS="$LDFLAGS $pmix_hwloc_LDFLAGS"
             fi
         fi
+      ])
+    else
+      pmix_cv_hwloc_support=0
     fi
 
-    if test ! -z "$with_hwloc" && test "$with_hwloc" != "no" && test "$pmix_hwloc_support" != "1"; then
+    if test ! -z "$with_hwloc" && test "$with_hwloc" != "no" && test "$pmix_cv_hwloc_support" != "1"; then
         AC_MSG_WARN([HWLOC SUPPORT REQUESTED AND NOT FOUND])
         AC_MSG_ERROR([CANNOT CONTINUE])
     fi
 
-    if test $pmix_hwloc_support = "1"; then
+    if test $pmix_cv_hwloc_support = "1"; then
         AC_MSG_CHECKING([if external hwloc version is 1.5 or greater])
         AC_COMPILE_IFELSE(
               [AC_LANG_PROGRAM([[#include <hwloc.h>]],
@@ -96,13 +98,13 @@ AC_DEFUN([PMIX_HWLOC_CONFIG],[
     fi
 
     AC_MSG_CHECKING([will hwloc support be built])
-    if test "$pmix_hwloc_support" != "1"; then
+    if test "$pmix_cv_hwloc_support" != "1"; then
         AC_MSG_RESULT([no])
     else
         AC_MSG_RESULT([yes])
     fi
 
-    AC_DEFINE_UNQUOTED([PMIX_HAVE_HWLOC], [$pmix_hwloc_support],
+    AC_DEFINE_UNQUOTED([PMIX_HAVE_HWLOC], [$pmix_cv_hwloc_support],
                        [Whether or not we have hwloc support])
     PMIX_VAR_SCOPE_POP
 ])dnl
```

this issue has to be addressed in upcoming `PMIx 3.1.0` since it will be embedded in Open MPI 4.0.1

Note we still pass `-lhwloc` in embedded mode if Open MPI uses an external hwloc.
I honestly do not know whether this is mandatory or not.
For example, we do not pass `-levent` if Open MPI uses an external pmix and everything is fine.

Bottom line, this patch might be indeed an overkill, we could for example never append `-lhwloc` in embedded mode, and let the `hwloc` version check fail if `PMIx` is embedded into an other project without `hwloc`.

**Comment by rhc54** [comment]
?? OMPI is supposed to pass ```--with-hwloc=xxx``` to the PMIx configure line to avoid this precise problem. Perhaps it is missing this since we didn't use hwloc in PMIx for awhile? Let me poke at it.

**Comment by rhc54** [comment]
Oh, I think I see what you are driving at - what happens when they pass ```--with-hwloc=internal```? Yeah, that would be a problem, but the best solution really is for us to deal with that in the OPAL configure code. We need to detect it there and ensure we set ```--with-hwloc``` to the right thing for the pmix component. Let me talk with @jsquyres about the best solution.

**Comment by ggouaillardet** [comment]
`—with-hwloc=internal` could be a distinct issue.

I faced the issue __without__ any `—with-hwloc` option, and Open MPI defaults to the embedded hwloc since it is more recent than the OS hwloc (e.g. CentOS 7)

Anyway, in both cases we do not want to append `-lhwloc` to the embedded pmix `LIBS`

**Comment by rhc54** [comment]
Okay, thanks for clarifying. I think we have two distinct issues here:

- we need to ensure we are looking at the correct hwloc to match what OMPI is using
- we don't want to add hwloc to the PMIx LIBS in embedded mode

I think the latter likely needs a change in PMIx itself, but the former is likely the sole responsibility of the OPAL configure code.

**Comment by ggouaillardet** [comment]
As long as ompi does not pass `—with-hwloc=internal` to the embedded pmix, I think PMIx looks at the hwloc it should.

**Comment by rhc54** [comment]
I believe I have addressed this - doing some more checking with OMPI now. I found the following statement to be puzzling:

> config/pmix_setup_libevent.m4 assumes everything is there when building the embedded mode (which is fine since libevent is a requirement of PMIx).
The very same logic cannot be applied here since hwloc is optional by default.

I see no reason why we cannot just use the libevent configure logic here. If the builder specifically passes the ```--enable-embedded-hwloc``` flag, then so far as I'm concerned, they are guaranteeing its presence _and_ directing us to enable PMIx hwloc support. Thus, there is nothing for us to check and all we have to do is set the header.

Feel free to correct my thinking - I simply based the PR on that rationale.
